---
title: "Ancestral SNPs Retrieval and Cleaning"
author: "Ahmed Moustafa"
date: "`r format(Sys.time(), '%d %B, %Y %H:%M %z')`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This notebook is to fetch the details of the 10,000 SNPs defined by [NCBI's GRAF](https://github.com/ncbi/graf) as the 10,000 ancestral SNPs. Once fetched, the data is processed to filter out SNPs and retain only the biallelic ones based on the information retrieved. The resultant dataset provides a focused set of SNPs that can be used for various analyses related to ancestry and population genetics.

### Data

The [rsIDs of the 10,000 SNPs](data/fingerprinting_snps.txt) were extracted from GRAF's [G1000FpGeno](https://github.com/ncbi/graf/tree/master/data) collection using [PLINK](https://www.cog-genomics.org/plink/)'s `--recode` option.

### References
- [GRAF GitHub](https://github.com/ncbi/graf)
- [GRAF-pop: A Fast Distance-Based Method To Infer Subject Ancestry from Multiple Genotype Datasets Without Principal Components Analysis](https://pubmed.ncbi.nlm.nih.gov/31151998/)

## Load Required Libraries
The following libraries are needed for SNP data retrieval and transformation:

```{r load-packages, message=FALSE}
library(rsnps)
library(tidyverse)
library(tidylog)
library(tictoc)
```

## Load Data

Load a list of rsIDs (reference SNP identifiers) from a provided text file.

```{r load-data}
# Load the list of rsIDs from a text file
rsids = readLines("data/fingerprinting_snps.txt")
head(rsids)
length(rsids)
```

```{r global-variables}
batch_size = 1000
batch_number = 1
```

## Define Function to Fetch SNP Data

A function is defined to fetch SNP data for a batch of rsIDs from the NCBI database. The batch processing is done to avoid overwhelming the server with too many requests at once.

```{r define-the-fetch-function, message=FALSE}
# Define a function to fetch SNP data for a batch of rsIDs
fetch_batch_snp_data = function(batch) {
    tryCatch({
        print (paste0("Processing batch number: ", batch_number))
        batch_number <<- batch_number + 1
        tic()
        result = tibble(ncbi_snp_query(batch))
        toc()
        return (result)
    }, error = function(e) {
        message(paste0("Error fetching data: ", e$message))
        NULL
  })
}
```

The rsIDs are split into batches, and the SNP data for each batch is fetched using the previously defined function.

```{r fetch-the-snps-info, message=FALSE}
# Split rsIDs into batches of 1,000 and fetch the SNP data
snp_data = map_dfr(split(rsids, ceiling(seq_along(rsids) / batch_size)), fetch_batch_snp_data)
```

## Overview of the Fetched SNP Data

Here's a glimpse of the fetched SNP data:

```{r snps-data-overview}
glimpse(snp_data)
```

## Data Cleaning

Remove unnecessary columns and rename certain columns for clarity.

```{r snps-cleaning}
# Mainly removing the list of the MAF per population (last column)
snp_data2 = snp_data %>% select(-query, -maf_population) %>% rename(position = bp)
glimpse(snp_data2)
```

## Save Cleaned SNP Data to TSV

The cleaned SNP data is saved to a TSV file for further analysis or storage.

```{r save-the-snps-info}
write_tsv(snp_data2, "data/fingerprinting_snps.tsv")
```

## Convert SNP Data to BED Format

The SNP data is converted into BED format, which is commonly used in genomics to represent genomic regions.

```{r create-bed-format}
# Create BED format
bed_df = snp_data2 %>%
    transmute(
        chrom = chromosome,
        start = position - 1,
        end = position,
        rsid = rsid)
glimpse(bed_df)
```

## Save BED Formatted Data

The BED formatted data is saved to a file.

```{r save-the-bed-format}
write_tsv(bed_df, "data/fingerprinting_snps.bed", col_names = FALSE)
```

## Filter for Bi-Allelic SNPs

Filter the SNP data to include only biallelic variants (SNPs with two alleles).

```{r rows of remove-multiallelic snps}
snp_data3 = snp_data2 %>% filter(!is.na(minor))
glimpse(snp_data3)
```

## Save Bi-Allelic SNP Data to TSV

The bi-allelic SNP data is saved to a TSV file.

```{r save-filtered-snps}
write_tsv(snp_data3, "data/fingerprinting_snps_biallelic.tsv")
```

## Convert Bi-Allelic SNP Data to BED Format

The bi-allelic SNP data is converted into BED format.

```{r convert-filtered-snps-to-bed}
bed_df2 = snp_data3 %>%
    transmute(
        chrom = chromosome,
        start = position - 1,
        end = position,
        rsid = rsid)
glimpse(bed_df2)
```

## Save Bi-Allelic BED Formatted Data

The BED formatted data for bi-allelic SNPs is saved to a file.

```{r save-the-converted-bed}
write_tsv(bed_df2, "data/fingerprinting_snps_biallelic.bed", col_names = FALSE)
```

## Session Info

```{r display-session-info}
sessionInfo()
```
